src/redisearch_rs/c_entrypoint/iterators_ffi/src/wrapper.rs

```diff
@@ -0,0 +1,128 @@
+//! A wrapper around a Rust iterator that allows the existing C code to invoke it
+//! as if it were a C iterator conforming to the existing C iterator conventions.
+use ffi::{
+    IteratorStatus, IteratorStatus_ITERATOR_EOF, IteratorStatus_ITERATOR_NOTFOUND,
+    IteratorStatus_ITERATOR_OK, IteratorStatus_ITERATOR_TIMEOUT, QueryIterator, ValidateStatus,
+    ValidateStatus_VALIDATE_ABORTED, ValidateStatus_VALIDATE_MOVED, ValidateStatus_VALIDATE_OK,
+    t_docId,
+};
+use inverted_index::RSIndexResult;
+use rqe_iterators::{RQEIterator, RQEIteratorError, RQEValidateStatus, SkipToOutcome};
+use std::ptr;
+
+#[repr(C)]
+pub struct RQEIteratorWrapper<E> {
+    // The iterator header.
+    // It *must* appear first for C-Rust interoperability to work as expected.
+    pub base: QueryIterator,
+    pub inner: E,
+}
+
+/// Convert a type-erased "base" iterator into a wrapper around a specific Rust iterator type.
+pub(crate) unsafe fn from_base<'index, I: RQEIterator<'index>>(
+    base: *mut QueryIterator,
+) -> &'index mut RQEIteratorWrapper<I> {
+    unsafe { &mut *(base as *mut RQEIteratorWrapper<I>) }
```

Comment:
personal preference: since the operations are all left-to-right instead of mixing postfix (`as *mut _` and prefix `&mut *(..)`)
```suggestion
    unsafe { base.cast::<RQEIteratorWrapper<I>>().as_mut() }
```

Reply:
I put this together with the goal of testing the integration—the code must absolutely be cleaned up before being viable for merging 😁 
Even before styling, the lack of `SAFETY` comments and documentation is a big no no.